[FEAT] question 기능 추가

hyeonic opened the pull request:
question 관련 기능들을 작성하였습니다. repository와 service layer는 간단한 테스트 코드도 작성하였습니다. 아직 세세한 validation이나 예외 처리 부분이 미숙하여 우선 입력이 성공하였을 때만 고려하여 작성하였습니다. 후에 좀 더 공부한 후 mock 을 활용한 controller 테스트를 진행할 예정입니다.

hyeonic committed:
feat: user 초기값 세팅

테스트를 위한 user 초기값을 세팅한다.

hyeonic committed:
feat: question 기능 추가

question repository 및 service 테스트 코드 작성
controller 구현
controller 테스트는 좀 더 공부한 후 작성할 예정
추가적인 validaiton은 아직 명시하지 않음

her0807 commented on src/main/resources/application.yml:
```diff
@@ -4,6 +4,8 @@ spring:
     username: sa
     password:
     driver-class-name: org.h2.Driver
+    data: classpath:data.sql
+    initialization-mode: always
```
확인했습니다

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/QuestionController.java:
```diff
@@ -0,0 +1,73 @@
+package com.skhuedin.skhuedin.controller;
+
+import com.skhuedin.skhuedin.domain.Question;
+import com.skhuedin.skhuedin.domain.User;
+import com.skhuedin.skhuedin.dto.question.QuestionMainResponseDto;
+import com.skhuedin.skhuedin.dto.question.QuestionSaveRequestDto;
+import com.skhuedin.skhuedin.service.QuestionService;
+import com.skhuedin.skhuedin.service.UserService;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+
```
패키지 별로 나누어 개행을 넣어주시면 조금 더 가독성이 좋아질 것 같습니다!

hyeonic replied:
최대한 인텔리제이가 자동으로 추가한 그대로 반영하였습니다. 후에 자체적으로 수정할 경우 코드의 충돌이 빈번해질 것 같아서 최대한 인텔리제이가 자동으로 추가된 코드를 유지하는 것이 좋을 것 같아요!

her0807 commented on src/main/java/com/skhuedin/skhuedin/controller/QuestionController.java:
```diff
@@ -0,0 +1,73 @@
+        return new ResponseEntity<>(new Result(questions), HttpStatus.OK);
+    }
+
+    @Data
+    @AllArgsConstructor
+    class Result<T> {
+        private T data;
+    }
```
```
new Result(questions)
```
위 처럼 객체 값을 담아서 엔티티에 저장하려고 만드신건가요?! 이 부분은 다른 컨트롤러에서도 중복적으로 사용가능성이 있으니 이너클래스가 아닌 따로 빼도 좋을 것 같아요

hyeonic replied:
네 맞습니다. json으로 전달 될 때 
```
{
    data: {
    
    }
}
```
이런식으로 감싸서 전달될 것 같아요!

her0807 commented on src/main/java/com/skhuedin/skhuedin/domain/User.java:
```diff
@@ -17,6 +18,7 @@
 @Entity
 @Getter
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@EqualsAndHashCode
```
추가하신 의도가 궁금합니다!

hyeonic replied:
테스트 코드 작성 시 User의 동등성과 동일성 비교를 위해 추가하였는데 현재 사용이 안되고 있어 삭제하였습니다.

her0807 commented on src/main/java/com/skhuedin/skhuedin/dto/question/QuestionMainResponseDto.java:
```diff
@@ -0,0 +1,35 @@
+    private UserMainResponseDto writerUser;
+    private String title;
+    private String content;
+    private Boolean status;
+    private Boolean fix;
```
## 논의 필요
fix를 여러개 하면 고정시킨 순서대로 or 작성한 순서대로 보여지게 할텐데 , 
고정시킨 순서대로 하게 된다면 필드를 추가하거나 불린 타입으로 하는 걸 조금 더 고려해봐야할 것 같습니다!

hyeonic replied:
네 우선 반영한 후 충분한 논의를 통하여 반영하면 좋을 것 같아요

her0807 commented on src/main/java/com/skhuedin/skhuedin/dto/question/QuestionSaveRequestDto.java:
```diff
@@ -0,0 +1,42 @@
+@Setter
+@NoArgsConstructor
+@ToString
+public class QuestionSaveRequestDto {
+
+    private String title;
+    private String content;
+    private User targetUser;
+    private User writerUser;
+    private Boolean status;
+    private Boolean fix;
+
+    @Builder
+    public QuestionSaveRequestDto(String title, String content, Boolean status, Boolean fix) {
+        this.title = title;
+        this.content = content;
+        this.status = status;
+        this.fix = fix;
+    }
+
+    public Question toEntity() {
+        return Question.builder()
+                .title(title)
+                .content(content)
+                .targetUser(targetUser)
+                .writerUser(writerUser)
+                .status(status)
+                .fix(fix)
+                .build();
+    }
+}
```
# [MapStruct 사용하기](https://www.skyer9.pe.kr/wordpress/?p=1596)
# [DTO 쉽게 만들기](https://huisam.tistory.com/entry/mapStruct)

기현님! DTO를 만들기위해 찾아보던 중 MapStruct 를 이용하면 DTO – Entity 간 매핑을 간편하게 할 수 있고, 더불어 Entity 업데이트 코드를 매우 간단하게 줄일 수 있는 것을 발견했습니다 ㅎㅎ 지금 구조도 좋은데 읽어보시면 좋을 것 같아 링크 남깁니다!

hyeonic replied:
mapstruct 사용도 좋지만 아직 확실히 되지 않는 api 스펙으로 인하여 dto 수정이 자주 야기될 것이 우려됩니다. 어느정도 프론트와 상의 후 정해지면 추가를 고려해보도록 할게요!

her0807 commented on src/test/java/com/skhuedin/skhuedin/repository/QuestionRepositoryTest.java:
```diff
@@ -0,0 +1,119 @@
+        );
+
+        // then
+        assertAll(() -> {
+            assertEquals(saveQuestion.getTitle(), findQuestion.getTitle());
+            assertEquals(saveQuestion.getContent(), findQuestion.getContent());
+            assertEquals(saveQuestion.getTargetUser(), findQuestion.getTargetUser());
+            assertEquals(saveQuestion.getWriterUser(), findQuestion.getWriterUser());
+            assertEquals(saveQuestion.getView(), findQuestion.getView());
+            assertEquals(saveQuestion.getStatus(), findQuestion.getStatus());
+            assertEquals(saveQuestion.getFix(), findQuestion.getFix());
+        });
+    }
```
중복되는 코드는 메소드로 따로 빼서 사용하시면 어떨까요? (이후에 계속 중복이 된다면..!)(물론 두번 밖에 사용안해서 안하셔도 될 것 같기도 해요..!)

hyeonic replied:
위 코드는 잘 못된 사용 방법을 가진 테스트 코드 였어요.

```java
assertAll(() -> {
            assertEquals(saveQuestion.getTitle(), findQuestion.getTitle());
            assertEquals(saveQuestion.getContent(), findQuestion.getContent());
            assertEquals(saveQuestion.getTargetUser(), findQuestion.getTargetUser());
            assertEquals(saveQuestion.getWriterUser(), findQuestion.getWriterUser());
            assertEquals(saveQuestion.getView(), findQuestion.getView());
            assertEquals(saveQuestion.getStatus(), findQuestion.getStatus());
            assertEquals(saveQuestion.getFix(), findQuestion.getFix());
        });
```

이것이 아니라 
```
assertAll(
                () -> assertEquals(question.getTitle(), saveQuestion.getTitle()),
                () -> assertEquals(question.getContent(), saveQuestion.getContent()),
                () -> assertEquals(question.getTargetUser(), saveQuestion.getTargetUser()),
                () -> assertEquals(question.getWriterUser(), saveQuestion.getWriterUser()),
                () -> assertEquals(question.getView(), saveQuestion.getView()),
                () -> assertEquals(question.getStatus(), saveQuestion.getStatus()),
                () -> assertEquals(question.getFix(), saveQuestion.getFix())
        );
```
이렇게 사용하는 방법이 맞는 방법이라고 합니다. 우선 간단한 테스트 코드 작성 후 후에 지속적으로 수정하겠습니다.

her0807 reviewed (approved):
question repository 및 service 테스트 코드 작성 깔끔하게 잘하신 것 같아요 ~ 리뷰 확인 부탁드려요
 validaiton 은 이슈 등록하신 부분에 의견 적어놓았습니다.! 오늘도 수고 많으셨어요

hyeonic committed:
feat: Question, Comment 기능 추가

Question과 Comment의 기능을 구현하였다.
service와 controller layer에 필요한 reqeusetDto와 responseDto를 추가
후에 비즈니스 요구사항이 변경되면 맞춰 수정할 예정
각 repository와 service에 대한 테스트 코드를 간단히 구현
